feat(middleware)!: forbid response body

feugy opened the pull request:
### What's in there?

As part of the Edge Functions General Availability roadmap, it has been decided to only support the following uses cases in Next.js' middleware:
- rewrite the URL (`x-middleware-rewrite` response header)
- redirect to another URL (`Location` response header)
- pass on to the next piece in the request pipeline (`x-middleware-next` response header)

1. during development, a warning on console tells developers when they are returning a response (either with `Response` or `NextResponse`).
2. at build time, this warning becomes an error.
3. at run time, returning a response body will trigger a 500 HTTP error with a JSON payload containing the detailed error.

All returned/thrown errors contain a link to the documentation.

This is a breaking feature compared to the _beta_ middleware implementation.
It also removes `NextResponse.json()`.

### How to try it?

The new error is covered with:
- integration tests: 
  - runtime behavior: `HEADLESS=true yarn jest test/integration/middleware/core`
  - build behavior : `HEADLESS=true yarn jest test/integration/middleware/build-errors`
  - development behavior: `HEADLESS=true yarn jest test/development/middleware-warnings`

### Notes to reviewers

The limitation happens in next's web adapter. The initial implementation was to check `response.body` existence, but it turns out [`Response.redirect()`](https://developer.mozilla.org/en-US/docs/Web/API/Response/redirect) may return an empty stream. Hence why the proposed implementation specifically looks at response headers.

Because this is a breaking change, I had to adjust several tests cases, previously returning JSON/stream/text bodies. When relevant, these middlewares are returning data using response headers.

About DevEx: relying on AST analysis to detect forbidden use cases is not as good as running the code.
Such cases are easy to detect:
```js
new Response('a text value')
new Response(JSON.stringify({ /* whatever */ })
```
But these are false-positive cases:
```js
function returnNull() { return null }
new Response(returnNull())

function doesNothing() {}
new Response(doesNothing())
```
However, I see no good reasons to let user build an middleware with the cases above, hence why the build will fail, even if _technically speaking_, they are not setting the response body.

feugy committed:
feat(middleware)\!: forbid response body

nkzawa commented on test/integration/middleware/core/test/index.test.js:
```diff
@@ -161,67 +163,34 @@ function urlTests(_log, locale = '') {
-      `${locale}/urls/relative-url`
-    )
-    const json = await res.json()
-    expect(json.error.message).toContain('Invalid URL')
```
Am I correct this is the case that error is thrown on a middleware invocation? I wonder what happens now after this PR.

feugy replied:
_discussed online_

@nkzawa correctly spot some tests that I should have not removed. Restoring them now.

feugy committed:
chore: restores URL legit test cases

feugy committed:
test: adds e2e tests for middleware body restriction

feugy committed:
doc: updates middleware and next/server pages.

feugy committed:
feat!: removes NextResponse.json() function

feugy committed:
feat: warns for middleware body response in development, fails during build

ijjk committed:
Update root component handling (#36781)

remcohaszing committed:
Add final newline to package-lock.json (#36813)

This aligns the output with what npm outputs. Without this change,
Next.js causes unwanted changes in package-lock.json.

## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [x] Make sure the linting passes by running `yarn lint`

Commit:
Fixing Tina + Adding a demo URL (#35480)

## Documentation / Examples

- Updating Tina Demo URL
- Adding Tina to blog-starter example
- Adding Tina to Basic features section
- Updating Tina to actually work again

shuding committed:
Move FlightManifestPlugin to server compilers (#36810)

* move FlightManifestPlugin to server compilers

* revert loader condition

* fix module id

* fix test and refactor

Co-authored-by: kodiakhq[bot] <49736102+kodiakhq[bot]@users.noreply.github.com>

ijjk committed:
Update moduleNameMapper jest config and remove extra deps (#36787)

* Update moduleNameMapper jest config and remove extra deps

* move svg mock to its own entry for easier overriding

* remove package script

feugy committed:
refactor(middleware): be as strict during build as in dev
Leverages Webpack builtin errors for better reporting

huozhi committed:
Wait for shell resolve with gIP is customized in react 18 (#36792)

When getInitialProps is customized with react 18, since gIP requires to return `html` as doc property which could be used by  user-land customization, we do blocking-rendering there and passdown the `html` to document

Fixes #36675
Closes #36419

feugy committed:
docs: improves education about the middleware body limitation

feugy commented:
Closing in favor of https://github.com/vercel/next.js/pull/36835
